File: native/android/runtime/src/main/kotlin/com/pegasusheavy/angularmobile/ViewFactory.kt
package com.pegasusheavy.angularmobile

import android.content.Context
import android.graphics.Color
import android.graphics.Typeface
import android.graphics.drawable.GradientDrawable
import android.text.InputType
import android.util.TypedValue
import android.view.Gravity
import android.view.View
import android.view.ViewGroup
import android.widget.*
import androidx.appcompat.widget.SwitchCompat
import androidx.core.widget.NestedScrollView
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import org.json.JSONObject

/**
 * View Factory
 *
 * Creates and updates native Android views based on commands from JavaScript.
 */
class ViewFactory(
    private val context: Context,
    private val viewRegistry: ViewRegistry
) {
    /**
     * Create a native view of the specified type
     */
    fun createView(viewType: String, viewId: String, props: JSONObject): View? {
        val view = when (viewType) {
            "View", "android.view.ViewGroup" -> createViewGroup(props)
            "Text", "android.widget.TextView" -> createTextView(props)
            "Image", "android.widget.ImageView" -> createImageView(props)
            "TextInput", "android.widget.EditText" -> createEditText(props)
            "Button", "android.widget.Button" -> createButton(props)
            "ScrollView", "android.widget.ScrollView" -> createScrollView(props)
            "Switch", "android.widget.Switch" -> createSwitch(props)
            "Slider", "android.widget.SeekBar" -> createSeekBar(props)
            "ActivityIndicator", "android.widget.ProgressBar" -> createProgressBar(props)
            "FlatList", "androidx.recyclerview.widget.RecyclerView" -> createRecyclerView(props)
            "WebView", "android.webkit.WebView" -> createWebView(props)
            "Modal", "android.app.Dialog" -> createModalContainer(props)
            else -> createViewGroup(props) // Default to ViewGroup
        }

        view?.let {
            applyCommonProps(it, props)
        }

        return view
    }

    /**
     * Update an existing view with new props
     */
    fun updateView(view: View, props: JSONObject) {
        applyCommonProps(view, props)

        when (view) {
            is TextView -> updateTextView(view, props)
            is ImageView -> updateImageView(view, props)
            is EditText -> updateEditText(view, props)
            is Button -> updateButton(view, props)
            is SwitchCompat -> updateSwitch(view, props)
            is SeekBar -> updateSeekBar(view, props)
            is ProgressBar -> updateProgressBar(view, props)
        }
    }

    // ==================== View Creators ====================

    private fun createViewGroup(props: JSONObject): ViewGroup {
        return FrameLayout(context).apply {
            layoutParams = ViewGroup.LayoutParams(
                ViewGroup.LayoutParams.MATCH_PARENT,
                ViewGroup.LayoutParams.WRAP_CONTENT
            )
        }
    }

    private fun createTextView(props: JSONObject): TextView {
        return TextView(context).apply {
            text = props.optString("text", "")
            updateTextStyles(this, props)
        }
    }

    private fun createImageView(props: JSONObject): ImageView {
        return ImageView(context).apply {
            scaleType = ImageView.ScaleType.CENTER_CROP
            updateImageView(this, props)
        }
    }

    private fun createEditText(props: JSONObject): EditText {
        return EditText(context).apply {
            hint = props.optString("placeholder", "")
            setText(props.optString("value", ""))
            updateEditText(this, props)
        }
    }

    private fun createButton(props: JSONObject): Button {
        return Button(context).apply {
            text = props.optString("title", "")
            isEnabled = !props.optBoolean("disabled", false)
        }
    }

    private fun createScrollView(props: JSONObject): NestedScrollView {
        return NestedScrollView(context).apply {
            isFillViewport = true
            isNestedScrollingEnabled = true
        }
    }

    private fun createSwitch(props: JSONObject): SwitchCompat {
        return SwitchCompat(context).apply {
            isChecked = props.optBoolean("value", false)
            isEnabled = !props.optBoolean("disabled", false)
        }
    }

    private fun createSeekBar(props: JSONObject): SeekBar {
        return SeekBar(context).apply {
            max = ((props.optDouble("maximumValue", 1.0) - props.optDouble("minimumValue", 0.0)) * 100).toInt()
            progress = ((props.optDouble("value", 0.0) - props.optDouble("minimumValue", 0.0)) * 100).toInt()
            isEnabled = !props.optBoolean("disabled", false)
        }
    }

    private fun createProgressBar(props: JSONObject): ProgressBar {
        val style = if (props.optString("size") == "large") {
            android.R.attr.progressBarStyleLarge
        } else {
            android.R.attr.progressBarStyle
        }
        return ProgressBar(context, null, style).apply {
            isIndeterminate = props.optBoolean("animating", true)
        }
    }

    private fun createRecyclerView(props: JSONObject): RecyclerView {
        return RecyclerView(context).apply {
            layoutManager = LinearLayoutManager(context)
            setHasFixedSize(true)
        }
    }

    private fun createWebView(props: JSONObject): android.webkit.WebView {
        return android.webkit.WebView(context).apply {
            settings.javaScriptEnabled = props.optBoolean("javaScriptEnabled", true)
            props.optString("source")?.let { url ->
                if (url.isNotEmpty()) loadUrl(url)
            }
        }
    }

    private fun createModalContainer(props: JSONObject): FrameLayout {
        return FrameLayout(context).apply {
            setBackgroundColor(Color.parseColor("#80000000"))
        }
    }

    // ==================== View Updaters ====================

    private fun updateTextView(view: TextView, props: JSONObject) {
        props.optString("text")?.let { view.text = it }
        updateTextStyles(view, props)
    }

    private fun updateImageView(view: ImageView, props: JSONObject) {
        // Image loading would typically use Glide/Coil
        props.optJSONObject("source")?.let { source ->
            source.optString("uri")?.let { uri ->
                // TODO: Load image from URI using image loading library
            }
        }

        when (props.optString("resizeMode")) {
            "cover" -> view.scaleType = ImageView.ScaleType.CENTER_CROP
            "contain" -> view.scaleType = ImageView.ScaleType.FIT_CENTER
            "stretch" -> view.scaleType = ImageView.ScaleType.FIT_XY
            "center" -> view.scaleType = ImageView.ScaleType.CENTER
        }
    }

    private fun updateEditText(view: EditText, props: JSONObject) {
        props.optString("placeholder")?.let { view.hint = it }
        props.optString("value")?.let {
            if (view.text.toString() != it) {
                view.setText(it)
            }
        }

        view.inputType = when {
            props.optBoolean("secureTextEntry", false) ->
                InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_PASSWORD
            props.optString("keyboardType") == "numeric" ->
                InputType.TYPE_CLASS_NUMBER
            props.optString("keyboardType") == "email-address" ->
                InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_VARIATION_EMAIL_ADDRESS
            props.optBoolean("multiline", false) ->
                InputType.TYPE_CLASS_TEXT or InputType.TYPE_TEXT_FLAG_MULTI_LINE
            else -> InputType.TYPE_CLASS_TEXT
        }

        view.isEnabled = props.optBoolean("editable", true)
    }

    private fun updateButton(view: Button, props: JSONObject) {
        props.optString("title")?.let { view.text = it }
        view.isEnabled = !props.optBoolean("disabled", false)
    }

    private fun updateSwitch(view: SwitchCompat, props: JSONObject) {
        view.isChecked = props.optBoolean("value", false)
        view.isEnabled = !props.optBoolean("disabled", false)
    }

    private fun updateSeekBar(view: SeekBar, props: JSONObject) {
        val min = props.optDouble("minimumValue", 0.0)
        val max = props.optDouble("maximumValue", 1.0)
        view.max = ((max - min) * 100).toInt()
        view.progress = ((props.optDouble("value", 0.0) - min) * 100).toInt()
        view.isEnabled = !props.optBoolean("disabled", false)
    }

    private fun updateProgressBar(view: ProgressBar, props: JSONObject) {
        view.isIndeterminate = props.optBoolean("animating", true)
        if (view.visibility == View.VISIBLE && props.optBoolean("hidesWhenStopped", true) && !props.optBoolean("animating", true)) {
            view.visibility = View.INVISIBLE
        }
    }

    // ==================== Style Application ====================

    private fun applyCommonProps(view: View, props: JSONObject) {
        val style = props.optJSONObject("style") ?: return

        // Dimensions
        val width = parseDimension(style.opt("width"))
        val height = parseDimension(style.opt("height"))

        view.layoutParams = (view.layoutParams ?: ViewGroup.MarginLayoutParams(
            width ?: ViewGroup.LayoutParams.WRAP_CONTENT,
            height ?: ViewGroup.LayoutParams.WRAP_CONTENT
        )).apply {
            if (this is ViewGroup.MarginLayoutParams) {
                leftMargin = dpToPx(style.optDouble("marginLeft", style.optDouble("margin", 0.0)))
                rightMargin = dpToPx(style.optDouble("marginRight", style.optDouble("margin", 0.0)))
                topMargin = dpToPx(style.optDouble("marginTop", style.optDouble("margin", 0.0)))
                bottomMargin = dpToPx(style.optDouble("marginBottom", style.optDouble("margin", 0.0)))
            }
        }

        // Padding
        view.setPadding(
            dpToPx(style.optDouble("paddingLeft", style.optDouble("padding", 0.0))),
            dpToPx(style.optDouble("paddingTop", style.optDouble("padding", 0.0))),
            dpToPx(style.optDouble("paddingRight", style.optDouble("padding", 0.0))),
            dpToPx(style.optDouble("paddingBottom", style.optDouble("padding", 0.0)))
        )

        // Background
        style.optString("backgroundColor")?.let { color ->
            if (color.isNotEmpty()) {
                try {
                    val bgDrawable = GradientDrawable().apply {
                        setColor(Color.parseColor(color))
                        cornerRadius = dpToPx(style.optDouble("borderRadius", 0.0)).toFloat()

                        val borderWidth = dpToPx(style.optDouble("borderWidth", 0.0))
                        if (borderWidth > 0) {
                            setStroke(borderWidth, Color.parseColor(
                                style.optString("borderColor", "#000000")
                            ))
                        }
                    }
                    view.background = bgDrawable
                } catch (e: Exception) {
                    // Invalid color
                }
            }
        }

        // Opacity
        view.alpha = style.optDouble("opacity", 1.0).toFloat()

        // Visibility
        when (style.optString("display")) {
            "none" -> view.visibility = View.GONE
            else -> view.visibility = View.VISIBLE
        }

        // Elevation (shadow)
        val elevation = style.optDouble("elevation", 0.0)
        if (elevation > 0) {
            view.elevation = dpToPx(elevation).toFloat()
        }
    }

    private fun updateTextStyles(view: TextView, props: JSONObject) {
        val style = props.optJSONObject("style") ?: props

        // Font size
        style.optDouble("fontSize", -1.0).let { size ->
            if (size > 0) {
                view.setTextSize(TypedValue.COMPLEX_UNIT_SP, size.toFloat())
            }
        }

        // Font color
        style.optString("color")?.let { color ->
            if (color.isNotEmpty()) {
                try {
                    view.setTextColor(Color.parseColor(color))
                } catch (e: Exception) { }
            }
        }

        // Font weight
        val fontWeight = style.optString("fontWeight", "normal")
        val fontStyle = style.optString("fontStyle", "normal")

        view.setTypeface(view.typeface, when {
            fontWeight == "bold" && fontStyle == "italic" -> Typeface.BOLD_ITALIC
            fontWeight == "bold" -> Typeface.BOLD
            fontStyle == "italic" -> Typeface.ITALIC
            else -> Typeface.NORMAL
        })

        // Text alignment
        view.gravity = when (style.optString("textAlign")) {
            "center" -> Gravity.CENTER
            "right", "end" -> Gravity.END
            "left", "start" -> Gravity.START
            else -> Gravity.START
        }

        // Line height
        style.optDouble("lineHeight", -1.0).let { height ->
            if (height > 0) {
                view.setLineSpacing(dpToPx(height - view.textSize / context.resources.displayMetrics.density).toFloat(), 1f)
            }
        }

        // Number of lines
        style.optInt("numberOfLines", 0).let { lines ->
            if (lines > 0) {
                view.maxLines = lines
                view.ellipsize = android.text.TextUtils.TruncateAt.END
            }
        }
    }

    // ==================== Utilities ====================

    private fun dpToPx(dp: Double): Int {
        return TypedValue.applyDimension(
            TypedValue.COMPLEX_UNIT_DIP,
            dp.toFloat(),
            context.resources.displayMetrics
        ).toInt()
    }

    private fun parseDimension(value: Any?): Int? {
        return when (value) {
            is Number -> dpToPx(value.toDouble())
            is String -> {
                when {
                    value.endsWith("%") -> null // Percentage needs parent measurement
                    value == "auto" -> ViewGroup.LayoutParams.WRAP_CONTENT
                    else -> value.toDoubleOrNull()?.let { dpToPx(it) }
                }
            }
            else -> null
        }
    }
}

File: native/ios/AngularMobile/Sources/ViewFactory.swift
import Foundation
import UIKit
import WebKit

/// View Factory
///
/// Creates and updates native iOS views based on commands from JavaScript.
public class ViewFactory {

    // MARK: - Properties

    private let viewRegistry: ViewRegistry

    // MARK: - Initialization

    public init(viewRegistry: ViewRegistry) {
        self.viewRegistry = viewRegistry
    }

    // MARK: - View Creation

    /// Create a native view of the specified type
    public func createView(type: String, viewId: String, props: [String: Any]) -> UIView? {
        let view: UIView?

        switch type {
        case "View":
            view = createContainerView(props: props)
        case "Text":
            view = createLabel(props: props)
        case "Image":
            view = createImageView(props: props)
        case "TextInput":
            view = createTextField(props: props)
        case "Button":
            view = createButton(props: props)
        case "ScrollView":
            view = createScrollView(props: props)
        case "Switch":
            view = createSwitch(props: props)
        case "Slider":
            view = createSlider(props: props)
        case "ActivityIndicator":
            view = createActivityIndicator(props: props)
        case "WebView":
            view = createWebView(props: props)
        case "FlatList":
            view = createTableView(props: props)
        case "Modal":
            view = createModalContainer(props: props)
        default:
            view = createContainerView(props: props)
        }

        if let view = view {
            applyCommonProps(to: view, props: props)
        }

        return view
    }

    /// Update an existing view with new props
    public func updateView(_ view: UIView, props: [String: Any]) {
        applyCommonProps(to: view, props: props)

        switch view {
        case let label as UILabel:
            updateLabel(label, props: props)
        case let imageView as UIImageView:
            updateImageView(imageView, props: props)
        case let textField as UITextField:
            updateTextField(textField, props: props)
        case let button as UIButton:
            updateButton(button, props: props)
        case let switchView as UISwitch:
            updateSwitch(switchView, props: props)
        case let slider as UISlider:
            updateSlider(slider, props: props)
        case let indicator as UIActivityIndicatorView:
            updateActivityIndicator(indicator, props: props)
        default:
            break
        }
    }

    // MARK: - View Creators

    private func createContainerView(props: [String: Any]) -> UIView {
        let view = FlexboxView()
        if let style = props["style"] as? [String: Any] {
            view.applyFlexStyle(style)
        }
        return view
    }

    private func createLabel(props: [String: Any]) -> UILabel {
        let label = UILabel()
        label.text = props["text"] as? String ?? ""
        label.numberOfLines = 0
        updateLabelStyles(label, props: props)
        return label
    }

    private func createImageView(props: [String: Any]) -> UIImageView {
        let imageView = UIImageView()
        imageView.contentMode = .scaleAspectFill
        imageView.clipsToBounds = true
        updateImageView(imageView, props: props)
        return imageView
    }

    private func createTextField(props: [String: Any]) -> UITextField {
        let textField = UITextField()
        textField.placeholder = props["placeholder"] as? String
        textField.text = props["value"] as? String
        textField.borderStyle = .roundedRect
        updateTextField(textField, props: props)
        return textField
    }

    private func createButton(props: [String: Any]) -> UIButton {
        let button = UIButton(type: .system)
        button.setTitle(props["title"] as? String ?? "", for: .normal)
        button.isEnabled = !(props["disabled"] as? Bool ?? false)
        return button
    }

    private func createScrollView(props: [String: Any]) -> UIScrollView {
        let scrollView = UIScrollView()
        scrollView.showsVerticalScrollIndicator = props["showsVerticalScrollIndicator"] as? Bool ?? true
        scrollView.showsHorizontalScrollIndicator = props["showsHorizontalScrollIndicator"] as? Bool ?? false
        scrollView.bounces = props["bounces"] as? Bool ?? true
        return scrollView
    }

    private func createSwitch(props: [String: Any]) -> UISwitch {
        let switchView = UISwitch()
        switchView.isOn = props["value"] as? Bool ?? false
        switchView.isEnabled = !(props["disabled"] as? Bool ?? false)
        return switchView
    }

    private func createSlider(props: [String: Any]) -> UISlider {
        let slider = UISlider()
        slider.minimumValue = Float(props["minimumValue"] as? Double ?? 0)
        slider.maximumValue = Float(props["maximumValue"] as? Double ?? 1)
        slider.value = Float(props["value"] as? Double ?? 0)
        slider.isEnabled = !(props["disabled"] as? Bool ?? false)
        return slider
    }

    private func createActivityIndicator(props: [String: Any]) -> UIActivityIndicatorView {
        let size = props["size"] as? String ?? "small"
        let style: UIActivityIndicatorView.Style = size == "large" ? .large : .medium
        let indicator = UIActivityIndicatorView(style: style)

        if props["animating"] as? Bool ?? true {
            indicator.startAnimating()
        }

        indicator.hidesWhenStopped = props["hidesWhenStopped"] as? Bool ?? true

        return indicator
    }

    private func createWebView(props: [String: Any]) -> WKWebView {
        let configuration = WKWebViewConfiguration()
        configuration.allowsInlineMediaPlayback = true

        let webView = WKWebView(frame: .zero, configuration: configuration)

        if let source = props["source"] as? [String: Any],
           let urlString = source["uri"] as? String,
           let url = URL(string: urlString) {
            webView.load(URLRequest(url: url))
        }

        return webView
    }

    private func createTableView(props: [String: Any]) -> UITableView {
        let tableView = UITableView(frame: .zero, style: .plain)
        tableView.separatorStyle = .none
        return tableView
    }

    private func createModalContainer(props: [String: Any]) -> UIView {
        let view = UIView()
        view.backgroundColor = UIColor.black.withAlphaComponent(0.5)
        return view
    }

    // MARK: - View Updaters

    private func updateLabel(_ label: UILabel, props: [String: Any]) {
        if let text = props["text"] as? String {
            label.text = text
        }
        updateLabelStyles(label, props: props)
    }

    private func updateImageView(_ imageView: UIImageView, props: [String: Any]) {
        if let source = props["source"] as? [String: Any],
           let urlString = source["uri"] as? String,
           let url = URL(string: urlString) {
            // Note: In production, use SDWebImage, Kingfisher, or Nuke
            URLSession.shared.dataTask(with: url) { data, _, _ in
                if let data = data, let image = UIImage(data: data) {
                    DispatchQueue.main.async {
                        imageView.image = image
                    }
                }
            }.resume()
        }

        switch props["resizeMode"] as? String {
        case "cover":
            imageView.contentMode = .scaleAspectFill
        case "contain":
            imageView.contentMode = .scaleAspectFit
        case "stretch":
            imageView.contentMode = .scaleToFill
        case "center":
            imageView.contentMode = .center
        default:
            break
        }
    }

    private func updateTextField(_ textField: UITextField, props: [String: Any]) {
        if let placeholder = props["placeholder"] as? String {
            textField.placeholder = placeholder
        }
        if let value = props["value"] as? String, textField.text != value {
            textField.text = value
        }

        textField.isSecureTextEntry = props["secureTextEntry"] as? Bool ?? false
        textField.isEnabled = props["editable"] as? Bool ?? true

        switch props["keyboardType"] as? String {
        case "numeric":
            textField.keyboardType = .numberPad
        case "email-address":
            textField.keyboardType = .emailAddress
        case "phone-pad":
            textField.keyboardType = .phonePad
        default:
            textField.keyboardType = .default
        }
    }

    private func updateButton(_ button: UIButton, props: [String: Any]) {
        if let title = props["title"] as? String {
            button.setTitle(title, for: .normal)
        }
        button.isEnabled = !(props["disabled"] as? Bool ?? false)
    }

    private func updateSwitch(_ switchView: UISwitch, props: [String: Any]) {
        if let value = props["value"] as? Bool {
            switchView.setOn(value, animated: true)
        }
        switchView.isEnabled = !(props["disabled"] as? Bool ?? false)
    }

    private func updateSlider(_ slider: UISlider, props: [String: Any]) {
        if let min = props["minimumValue"] as? Double {
            slider.minimumValue = Float(min)
        }
        if let max = props["maximumValue"] as? Double {
            slider.maximumValue = Float(max)
        }
        if let value = props["value"] as? Double {
            slider.setValue(Float(value), animated: true)
        }
        slider.isEnabled = !(props["disabled"] as? Bool ?? false)
    }

    private func updateActivityIndicator(_ indicator: UIActivityIndicatorView, props: [String: Any]) {
        let animating = props["animating"] as? Bool ?? true
        if animating {
            indicator.startAnimating()
        } else {
            indicator.stopAnimating()
        }
    }

    // MARK: - Style Application

    private func applyCommonProps(to view: UIView, props: [String: Any]) {
        guard let style = props["style"] as? [String: Any] else { return }

        // Background color
        if let bgColor = style["backgroundColor"] as? String {
            view.backgroundColor = UIColor(hex: bgColor)
        }

        // Opacity
        if let opacity = style["opacity"] as? Double {
            view.alpha = CGFloat(opacity)
        }

        // Border
        if let borderWidth = style["borderWidth"] as? Double {
            view.layer.borderWidth = CGFloat(borderWidth)
        }
        if let borderColor = style["borderColor"] as? String {
            view.layer.borderColor = UIColor(hex: borderColor)?.cgColor
        }
        if let borderRadius = style["borderRadius"] as? Double {
            view.layer.cornerRadius = CGFloat(borderRadius)
            view.clipsToBounds = true
        }

        // Shadow
        if let shadowColor = style["shadowColor"] as? String {
            view.layer.shadowColor = UIColor(hex: shadowColor)?.cgColor
        }
        if let shadowOpacity = style["shadowOpacity"] as? Double {
            view.layer.shadowOpacity = Float(shadowOpacity)
        }
        if let shadowRadius = style["shadowRadius"] as? Double {
            view.layer.shadowRadius = CGFloat(shadowRadius)
        }
        if let shadowOffset = style["shadowOffset"] as? [String: Double] {
            view.layer.shadowOffset = CGSize(
                width: shadowOffset["width"] ?? 0,
                height: shadowOffset["height"] ?? 0
            )
        }

        // Visibility
        if let display = style["display"] as? String, display == "none" {
            view.isHidden = true
        } else {
            view.isHidden = false
        }
    }

    private func updateLabelStyles(_ label: UILabel, props: [String: Any]) {
        let style = props["style"] as? [String: Any] ?? props

        // Font size
        if let fontSize = style["fontSize"] as? Double {
            label.font = label.font.withSize(CGFloat(fontSize))
        }

        // Font color
        if let color = style["color"] as? String {
            label.textColor = UIColor(hex: color)
        }

        // Font weight
        var fontWeight: UIFont.Weight = .regular
        if let weight = style["fontWeight"] as? String {
            switch weight {
            case "bold", "700", "800", "900":
                fontWeight = .bold
            case "600":
                fontWeight = .semibold
            case "500":
                fontWeight = .medium
            case "300":
                fontWeight = .light
            default:
                fontWeight = .regular
            }
        }

        let fontSize = style["fontSize"] as? Double ?? Double(label.font.pointSize)
        label.font = UIFont.systemFont(ofSize: CGFloat(fontSize), weight: fontWeight)

        // Font style
        if let fontStyle = style["fontStyle"] as? String, fontStyle == "italic" {
            if let descriptor = label.font.fontDescriptor.withSymbolicTraits(.traitItalic) {
                label.font = UIFont(descriptor: descriptor, size: label.font.pointSize)
            }
        }

        // Text alignment
        if let textAlign = style["textAlign"] as? String {
            switch textAlign {
            case "center":
                label.textAlignment = .center
            case "right", "end":
                label.textAlignment = .right
            default:
                label.textAlignment = .left
            }
        }

        // Number of lines
        if let numberOfLines = style["numberOfLines"] as? Int {
            label.numberOfLines = numberOfLines
        }
    }
}

// MARK: - UIColor Extension

extension UIColor {
    convenience init?(hex: String) {
        var hexSanitized = hex.trimmingCharacters(in: .whitespacesAndNewlines)
        hexSanitized = hexSanitized.replacingOccurrences(of: "#", with: "")

        var rgb: UInt64 = 0
        Scanner(string: hexSanitized).scanHexInt64(&rgb)

        let length = hexSanitized.count

        let r, g, b, a: CGFloat
        switch length {
        case 6:
            r = CGFloat((rgb & 0xFF0000) >> 16) / 255.0
            g = CGFloat((rgb & 0x00FF00) >> 8) / 255.0
            b = CGFloat(rgb & 0x0000FF) / 255.0
            a = 1.0
        case 8:
            r = CGFloat((rgb & 0xFF000000) >> 24) / 255.0
            g = CGFloat((rgb & 0x00FF0000) >> 16) / 255.0
            b = CGFloat((rgb & 0x0000FF00) >> 8) / 255.0
            a = CGFloat(rgb & 0x000000FF) / 255.0
        default:
            return nil
        }

        self.init(red: r, green: g, blue: b, alpha: a)
    }
}
